### ISSUE TITLE
Train loss explode MLA.ipynb with long seq_len

### ISSUE THREAD

**Issue by mw66** [Atulit23/ai_from_scratch#1]
MLA.ipynb 

If we change:
```
vocab_size = 320
seq_len = 2560
```
And move `inputs = torch.randint(...)` out of the epoch loop (i.e train the same data over and over again), you will see the training loss explode.

Not sure where is the bug.

**Comment by Atulit23** [comment]
Hey there, so this is not an exact implementation, I made some changes (refrained from using certain techniques), to optimized it such that it can run even if the memory you have is less. You can try making the model deeper, that is, increasing initialized values like d_model, n_layers, ffn_dim , d_rope etc. & you'll see some better performance. 

Lemme know if you'd like to know, what changes i made ;)

**Comment by mw66** [comment]
Well, it's a neutral network: in theory the back propagation should always work to make the loss smaller during training. That is why I suspect there is a bug some where.

**Comment by Atulit23** [comment]
No no you're correct, I've made some compromises in the model architecture so that I could run it on lower memory but because of that it does not do well for longer sequences. Like I have done just one single linear projection for KV compression instead of multilayered, this way it consumes less memory but results in higher loss & a couple others just like this.

**Comment by lmalololol** [comment]
Well, what did you change?
